src/gui/PresentationLaser.cpp

```diff
@@ -0,0 +1,62 @@
+#include "PresentationLaser.h"
+
+#include <cmath>
+
+static gboolean timeoutCallback(PresentationLaser* self) {
+    self->step();
+    return true;
+}
+
+PresentationLaser::PresentationLaser(GtkWidget* widget): widget(widget) {
+    this->timeoutId = g_timeout_add(DELTA, reinterpret_cast<GSourceFunc>(timeoutCallback), this);
+    g_object_ref(this->widget);
+}
+
+PresentationLaser::~PresentationLaser() {
+    if (this->timeoutId != -1) {
+        g_source_remove(this->timeoutId);
+    }
+    g_object_unref(this->widget);
+}
+
+void PresentationLaser::step() {
+    const double distX = this->target.x - this->x;
+    const double distY = this->target.y - this->y;
+
+    if (distX == 0.0 && distY == 0.0) {
+        return;
+    }
+
+    // The laser will move, so redraw the current position
+    const Rectangle<double>&& oldBounds = this->getDrawBounds();
+    gtk_widget_queue_draw_area(this->widget, oldBounds.x, oldBounds.y, oldBounds.width, oldBounds.height);
+
+    const double angle = std::atan2(distY, distX);
+    const double dx = std::abs(distX) < this->speed ? distX : speed * std::cos(angle);
+    const double dy = std::abs(distY) < this->speed ? distY : speed * std::sin(angle);
+    this->x += dx;
+    this->y += dy;
+
```

Comment:
If you limit xspeed and yspeed separately this can lead to a strange effect: If you move almost horizontally or almost vertically (only one direction faster than speed) then the laser will move at a steeper angle, rich that horizontal or vertical line and then move strictly horizontally or vertically.